src/solc_0.8/polygon/child/land/PolygonLandTunnelV2.sol

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.2;
+
+import "../../../common/fx-portal/FxBaseChildTunnelUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+import "../../../common/interfaces/IPolygonLand.sol";
+import "../../../common/interfaces/IERC721MandatoryTokenReceiver.sol";
+import "../../../common/BaseWithStorage/ERC2771Handler.sol";
+import "./PolygonLandBaseToken.sol";
+
+contract PolygonLandTunnelV2 is
+    FxBaseChildTunnelUpgradeable,
+    IERC721MandatoryTokenReceiver,
+    ERC2771Handler,
+    OwnableUpgradeable,
+    PausableUpgradeable
+{
+    IPolygonLand public childToken;
+    uint32 public maxGasLimitOnL1;
+    uint256 public maxAllowedQuads;
+    bool internal transferringToL1;
+
+    mapping(uint8 => uint32) public gasLimits;
+
+    event SetGasLimit(uint8 size, uint32 limit);
+    event SetMaxGasLimit(uint32 maxGasLimit);
+    event SetMaxAllowedQuads(uint256 maxQuads);
+    event Deposit(address indexed user, uint256 size, uint256 x, uint256 y, bytes data);
+    event Withdraw(address indexed user, uint256 size, uint256 x, uint256 y, bytes data);
+
+    function setMaxLimitOnL1(uint32 _maxGasLimit) external onlyOwner {
+        maxGasLimitOnL1 = _maxGasLimit;
+        emit SetMaxGasLimit(_maxGasLimit);
+    }
+
+    function setMaxAllowedQuads(uint256 _maxAllowedQuads) external onlyOwner {
+        require(_maxAllowedQuads > 0, "PolygonLandTunnel: max allowed value cannot be zero");
+        maxAllowedQuads = _maxAllowedQuads;
+        emit SetMaxAllowedQuads(_maxAllowedQuads);
+    }
+
+    function _setLimit(uint8 size, uint32 limit) internal {
```

Comment:
I'd organize the contract with public and external functions first, and internal in the end